### ISSUE TITLE
DOC: add example for using tensorflow

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#814]
#### Describe the issue linked to the documentation

We're not super clear about supporting any kind of model for our (currently existing reductions and postprocessing) mitigation techniques. This means people don't know they can use tensorflow with Fairlearn.

#### Suggest a potential alternative/fix

We should have a short example that shows how to do that. Perhaps having it as a `plot_tensorflow.py` under `examples` and linking to it from a user guide section makes sense. We've used `literalinclude` for this purpose in parts of the codebase. I am not entirely sure whether including tensorflow into requirements-dev.txt will cause issues, though.

The alternative: Just put a `code-block`, but then the code isn't run during the doc build and we're at risk of this code getting outdated. 

Related: #813

**Comment by riedgar-ms** [comment]
I'd worry that if we start adding tensorflow, pytorch etc. to our `requirements-dev.txt` (the only way to get code run in the documentation) we're going to discover that they have mutually incompatible dependencies (and possibly require different Python versions). We could point people at the `test_othermlpackages` directory, but that code isn't written as a teaching example :-(

**Comment by romanlutz** [comment]
Indeed. I would recommend adding a `code-block` to start with. It wouldn't get run during the doc build, so we'd avoid the issue you mention (which I mentioned in the description as well). As long as it's consistent with the code in the unit tests we can be confident it works.

**Comment by MiroDudik** [comment]
@romanlutz : I'm a bit confused, why should the example be named `plot_tensorflow.py`? If I understand the issue correctly, it's not just about plotting, but possibly other ways of linking the assessment and mitigation with tensorflow?

**Comment by romanlutz** [comment]
All example files need to be named something that starts with plot, it's required by Sphinx gallery.

**Comment by MiroDudik** [comment]
Ouch... that seems like an anti-pattern (but I see it in sklearn too). But we can deal with that another day...

**Comment by MiroDudik** [comment]
tagging @hannanabdul55

**Comment by riedgar-ms** [comment]
@MiroDudik It's like test files for `pytest` having to be named `test_<something>.py`. I'm not a huge fan, but know when I'm swimming against the tide.

**Comment by romanlutz** [comment]
@riedgar-ms is right, this isn't necessarily an anti-pattern in the way we've adopted it. It's just the naming convention adopted by sphinx-gallery because they store plots/pictures generated from these python files in the gallery and you can subsequently reference them. That's the purpose of sphinx-gallery (unlike the example notebooks that we've been using it for). I don't think they even allow us to configure that any other way.

**Comment by pritishaw1** [comment]
Hi @romanlutz, I would like to contribute a notebook for this issue. I have used the dataset shap.datasets.adult() to train a neural network model, using TensorFlow and Keras. After this, I used the fairlearn library that displays a dashboard for testing the fairness of the model built. This could serve as the perfect example to use TensorFlow, along with the fairlearn library. I will create a pull request, once you approve.

Thank You,
Priti

**Comment by adrinjalali** [comment]
@pritishaw1 please note that it should be in the form of an example python file, and not a notebook. Check https://github.com/fairlearn/fairlearn/tree/main/examples for other examples.

I'm not sure what you mean by dashboard, but the dashboard is deprecated and you can use the alternative plotting functionality we have now.

I'm also not sure how you have use fairlearn with TF, but you probably should use scikeras for this example, since it can be used nicely by others as well.

**Comment by pritishaw1** [comment]
@adrinjalali thanks for the response.

After reading through your comment, I am in a dilemma that whether the solution that I have would do justice to the issue.

Could you please help me a little more, by briefly explaining what exactly is expected as a solution to this issue?

**Comment by adrinjalali** [comment]
Your scenario sounds good, but on the implementation side, fairleanr expects a scikit-learn compatible estimator, and the recommended way for people to wrap their TF models into a scikit-learn compatible estimator is by using https://github.com/adriangb/scikeras

And when it comes to visualization, please look at the recent updates such as https://github.com/fairlearn/fairlearn/pull/806 , https://github.com/fairlearn/fairlearn/pull/770 , and https://github.com/fairlearn/fairlearn/pull/767 to see how you can visualize your results. Does this help?

**Comment by pritishaw1** [comment]
@adrinjalali can you please verify if this is correct? 

https://colab.research.google.com/drive/1olUuihBk_9JVH6lwA2AijKZFT_p61zU_?usp=sharing

I have implemented this in the jupyter notebook, where the dashboard is getting displayed, but as it is not supported with colab, it does not show up here. 

Instead, if this is not what's needed for the example using TensorFlow, then please do let me know how can I make a successful contribution to this issue.

**Comment by adrinjalali** [comment]
In that notebook, you're using the dashboard, and as I mentioned, you should instead use the matplotlib based plotting functionalities, and you can see examples of it in the three PRs I linked above. Your example should also not depend on SHAP, and should instead use our datasets module, or directly use `fetch_openlm` provided by scikit-learn.

Would it be maybe easier for you to start with one of the issues listed here? https://github.com/fairlearn/fairlearn/labels/good%20first%20issue

**Comment by pritishaw1** [comment]
Okay @adrinjalali, now I got it. I'll consider all the points that you have mentioned above, this time, and look out for its implementation.

I will also consider the other issues that you have listed above.

Thank you!

**Comment by pritishaw1** [comment]
Hello @adrinjalali,
I have implemented a notebook for this issue, following the points that you've mentioned above.

I have used the fetch_openlm from sklearn.datasets and you said, that fairlearn uses scikit-learn compatible estimator, I've used scikeras for wrapping the TensorFlow model.

Besides, I also went for matplotlib based plotting this time, and have used fairlearn.metric for that. 

Can you please verify, if this is what's needed for this issue?
Here's the link to it.
https://github.com/pritishaw1/Fairlearn_Tensorflow/blob/main/Fairlearn_plot_tensorflow.ipynb

Thank you!

**Comment by adrinjalali** [comment]
Yes, it looks much better, now you can add a story and comments around your code as you see in our `examples` folder. @MiroDudik and @hildeweerts would be probably better judges of the content and the story of the file.

**Comment by pritishaw1** [comment]
@adrinjalali Thanks for the review.
I will start with the story and the comments around the code and will reach out to the team soon, for their approval.

**Comment by romanlutz** [comment]
The part I don't quite understand yet is how this will work with the dependencies (as mentioned in the issue description).

**Comment by pritishaw1** [comment]
@romanlutz For this, we can go by adding a code-block for the issue, so that people know that they can use TensorFlow with fairlearn. 
And as you have recommended before that adopting this would ensure that the file does not get run during document build, we can overcome the issues of incompatible dependencies.

**Comment by romanlutz** [comment]
The code-block is a directive, though, so that would perhaps work in the user guide, but not in the `examples`. In the `examples` we have code that's actually executed, so that won't work without taking a (dev) dependency on tensorflow/keras. So I'd suggest having the 3-5 specific lines of code needed for plugging tensorflow/keras classifiers into our mitigation techniques as the `estimator` argument and put those into a `code-block` in the user guide. I suppose a subsection under `mitigation.rst` would make sense (I'm thinking 3.1, before the current 3.1 section). https://fairlearn.org/main/user_guide/mitigation.html

**Comment by pritishaw1** [comment]
Got it. So, as per your suggestions, I will go by adding some specific code lines for this, as a code-block in the user guide as subsection 3.1 under mitigation.rst, so that the dependencies are not disturbed. Here "fairlearn/docs/user_guide/mitigation.rst"

**Comment by romanlutz** [comment]
@fairlearn/fairlearn-maintainers does that sound right to you?
In any case, the code-block shouldn't include all the code that would go into an example, but rather the important lines only (should be no more than a handful or so). I believe your current example only uses Fairlearn in assessment, but not yet for mitigation. Assessment isn't really different for tensorflow models since it's just about the predictions/output, so I don't think we'd need to mention tensorflow in any way there. For mitigation we (might) need to plug that model into one of Fairlearn's mitigation techniques, so that's somewhat different. 

Out of curiosity I check the pytorch tutorials and it turns out they have an entirely separate repository for the tutorial build. I haven't quite figured out how they merge the generated docs with the generated docs from the basic doc build (for API reference etc.), but that would be another option (although much more involved, of course). I think we can keep this in mind for the future, although it's not really viable for this issue right now.

**Comment by adrinjalali** [comment]
yeah I'd be happy not to include tensorflow as a dependency. Missed that about this issue.

**Comment by pritishaw1** [comment]
@adrinjalali @romanlutz So, shall I proceed, and add some specific lines of the code, addressing this issue, in the form of a code-block, under the user guide section?
And from some specific lines, I mean lines between 37 to 70, along with some important imports from this https://github.com/pritishaw1/Fairlearn_Tensorflow/blob/main/plot_tensorflow.py

**Comment by romanlutz** [comment]
@pritishaw1 your example doesn't actually use our mitigation techniques yet. The classifier should be plugged into `ExponentiatedGradient` or `GridSearch` or `ThresholdOptimizer` (the last one only if we can get it to produce scores rather than binary output).

**Comment by pritishaw1** [comment]
@romanlutz Thanks for the details. I'll try using the mitigation techniques to it.
